src/main/java/in/org/projecteka/hiu/consent/ConsentService.java

```diff
@@ -114,32 +117,89 @@ public ConsentService(HiuProperties hiuProperties,
+                        var consentCreatedDate = (LocalDateTime) consent.get("dateCreated");
+                        if (consentCreatedDate.isAfter(now.minusMinutes(consentServiceProperties.getConsentRequestDelay()))) {
+                            return Mono.empty();
+                        }
+                        if (consentArtefactId == null) {
+                            continue;
+                        } else {
+                            return patientConsentRepository.getDataFlowParts(consentArtefactId).flatMap(dataFlowParts -> {
+                                DateRange dateRange = (DateRange) consent.get("dateRange");
+                                if (dataFlowParts.isEmpty()) {
+                                    var consentRequestId = consent.get("consentRequestId").toString();
+                                    return consentRepository.consentRequestForConsentRequestId(consentRequestId).flatMap(consentStatus -> {
+                                        if (consentStatus.equals(ConsentStatus.ERRORED) || consentStatus.equals(EXPIRED)) {
+                                            return buildConsentRequest(requesterId, hipId, dateRange.getFrom());
+                                        } else {
+                                            return Mono.empty();
+                                        }
+                                    });
+                                } else {
+                                    for (Map<String, Object> dataFlowPart : dataFlowParts) {
+                                        var latestResourceDate = (LocalDateTime) dataFlowPart.get("latestResourceDate");
+                                        var dataFlowStatus = dataFlowPart.get("status").toString();
+                                        if (HealthInfoStatus.valueOf(dataFlowStatus).equals(SUCCEEDED)) {
+                                            return latestResourceDate == null ?
+                                                    buildConsentRequest(requesterId, hipId, dateRange.getFrom()) :
+                                                    buildConsentRequest(requesterId, hipId, latestResourceDate);
+                                        }
+                                        if (HealthInfoStatus.valueOf(dataFlowStatus).equals(HealthInfoStatus.ERRORED)) {
+                                            return latestResourceDate != null ?
+                                                    buildConsentRequest(requesterId, hipId, latestResourceDate) :
+                                                    buildConsentRequest(requesterId, hipId, dateRange.getFrom());
+
+                                        }
+                                    }
+                                }
+                                return Mono.empty();
+                            });
+                        }
+                    }
+                    return Mono.empty();
```

Comment:
Can this be extracted out into a method?

Reply:
Tried but no luck, Need to pass lot of context to function